### PR TITLE
Expose direction_type string in navitia

### PR DESCRIPTION
**Pull request by vpassama**
Now this PR consists in adding a new field named direction_type in routes.txt NTFS file as we can see in the last update of the [documentation](https://github.com/prhod/navitia/blob/ntfs_doc_0.6/documentation/ntfs/ntfs_0.6.md).

I didn't add the tests for now but I corrected the little mistakes about line length (113 chars) in the source code.

Now the result looks like this : 

routes_direction_type

Outdated : 

Looking at the [documentation](https://github.com/CanalTP/navitia/blob/dev/documentation/ntfs/ntfs_0.5.md#routestxt-requis) about the routes.txt NTFS file, we can see the 'is_forward' required field.

This PR is just about exposing this information in the route API.

It is my first contribution in navitia and even if this is not huge, if something is wrong please tell me and I'll fix it.

This PR involves changes in navitia-proto, as you can see [here](https://github.com/CanalTP/navitia-proto/pull/19).

You can see an example of result in API : 

is_forward

### CODE REVIEW

File: source/ed/connectors/fusio_parser.h [hunk 1]
```diff
@@ -88,7 +88,7 @@ struct RouteFusioHandler : public GenericHandler {
     int ignored;
     void init(Data&);
     void handle_line(Data& data, const csv_row& line, bool is_first_line);
-    const std::vector<std::string> required_headers() const { return {"route_id", "route_name", "line_id"}; }
+    const std::vector<std::string> required_headers() const { return {"route_id", "route_name", "line_id", "is_forward"}; }
```
Comment (pbougue) [inline source/ed/connectors/fusio_parser.h]: line length: for readability we try to stay under 113 chars which is the limit in github

**Comment by antoine-de** [comment]
Hi @vpassama , 

I know it's in our documentation, but I don't think the `is_forward` field is the best way to handle this, I would prefer for the field to be named `direction_id` and to be a string.

In the GTFS specification, in trips.txt there is a `direction_id` field. Since there is no notion of line/route in the GTFS we create one `Line` per line in `routes.txt` and one `Route` per {line, direction_id} in `trips.txt` (https://github.com/CanalTP/navitia/blob/dev/source/ed/connectors/gtfs_parser.cpp#L919). We did one change to the GTFS specification on this though and consider this field a string, not a boolean (since it's a superset of the GTFS specification, we still implement the spec). This makes it possible for people feeding GTFS in navitia to create any number of `Route`s for a `Line`.

I think it would be more coherent to just store this GTFS `direction_id` field in the `Route` object, and have the same field in the NTFS.

I also don't see why this field should be required.

To sum it up:
the suggestion is to change this required bool field `is_forward` to an optional string field `direction_id`

What do you think about that @stifoon, @prhod ?

In a pure technical aspect, your PR is OK for me, thanks for the work !

I would just ask you to add some docker tests (I know this part is not well documented :wink: ).

Those tests are needed to test the Ed database handling. We wrap an Ed database inside a docker, load data into it, and export them with ed2nav to test the loaded data.
You need to change the files in fixture/ed/ntfs/ for your need, install docker (a recent version would be better), and compile the test with:
`make docker_test`

you need to change the file https://github.com/CanalTP/navitia/blob/dev/source/ed/docker_tests/ed_integration_tests.cpp to check that some routes have the wanted field.
Do not hesitate if you have questions on this.

Antoine

**Comment by prhod** [comment]
Hello @vpassama 
As discussed earlier with @antoine-de, we want to allow a string instead of the forward/backward boolean.
The field name will be "direction_type" to be consistent with the Netex specification (and not so far from the direction_id of the GTFS).
The recommanded values will be specified in the NTFS documentation, let start with  : 
- forward / backward
- clockwise / anticlockwise 
- outbound / inbound
  Is that OK with you ?
  Pascal

**Comment by vpassama** [comment]
Hi, that's ok for us. Thanks @prhod and @antoine-de. 

About the test @antoine-de, we will get a look at docker and if we have a problem we will ask you for some help. Have a nice day.

File: source/type/data.cpp [hunk 2]
```diff
@@ -63,7 +63,7 @@ namespace navitia { namespace type {
 
 wrong_version::~wrong_version() noexcept {}
 
-const unsigned int Data::data_version = 49; //< *INCREMENT* every time serialized data are modified
+const unsigned int Data::data_version = 50; //< *INCREMENT* every time serialized data are modified
```
Comment (pbougue) [inline source/type/data.cpp]: warning: you might need to re-increment that before merge as it is also incremented in another PR, especially if it is not merged in the 1.31 version.
(the important thing is that it is incremented at least once between 2 navitia version if needed)

Comment (vpassama) [inline source/type/data.cpp]: Thanks for pointing this out, I'll change it.

**Comment by prhod** [comment]
The NTFS 0.5 has been published, so the replacement of the is_forward field by the direction_type has been made in the NTFS 0.6 PR https://github.com/CanalTP/navitia/pull/1250

**Comment by vpassama** [comment]
I have an issue with make docker_test, maybe you can help me on this @antoine-de ?

```
DEBUG:requests.packages.urllib3.connectionpool:"POST /run/docker.sock/v1.12/containers/create HTTP/1.1" 201 90
INFO:docker_wrapper:docker id is 6901eac74667b4571ce87c3088dc1014b2d13c0c52fbfe36d637261845004313
INFO:docker_wrapper:starting the temporary docker
DEBUG:requests.packages.urllib3.connectionpool:"POST /run/docker.sock/v1.12/containers/6901eac74667b4571ce87c3088dc1014b2d13c0c52fbfe36d637261845004313/start HTTP/1.1" 204 0
DEBUG:requests.packages.urllib3.connectionpool:"GET /run/docker.sock/v1.12/containers/6901eac74667b4571ce87c3088dc1014b2d13c0c52fbfe36d637261845004313/json HTTP/1.1" 200 None
missing "=" after "postgresql://docker:docker@172.17.0.2/postgres" in connection info string

make[3]: *** [ed/docker_tests/ntfs_data.nav.lz4] Erreur 1
make[2]: *** [ed/docker_tests/CMakeFiles/datanav_files.dir/all] Erreur 2
make[1]: *** [CMakeFiles/docker_test.dir/rule] Erreur 2
make: *** [docker_test] Erreur 2
```

**Comment by antoine-de** [comment]
seems like a lib version problem. Can you come on irc ?

Did you `pip install -r navitia/source/eitri/requirement.txt` ?

File: source/ed/connectors/fusio_parser.h [hunk 3]
```diff
@@ -88,7 +88,9 @@ struct RouteFusioHandler : public GenericHandler {
     int ignored;
     void init(Data&);
     void handle_line(Data& data, const csv_row& line, bool is_first_line);
-    const std::vector<std::string> required_headers() const { return {"route_id", "route_name", "line_id"}; }
+    const std::vector<std::string> required_headers() const {
+        return {"route_id", "route_name", "line_id", "direction_type"};
```
Comment (antoine-de) [inline source/ed/connectors/fusio_parser.h]: I don't think the direction_type is required.

what do you think about this @prhod ?

File: source/sql/alembic/versions/14346346596e_add_direction_type_column_in_route_table.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+down_revision = '59f4456a029'
+
+from alembic import op
+import sqlalchemy as sa
+import geoalchemy2 as ga
+
+
+def upgrade():
+    op.add_column('route', sa.Column('direction_type', sa.TEXT(), nullable=True), schema='navitia')
+    op.drop_column('route', 'is_forward', schema='navitia')
```
Comment (antoine-de) [inline source/sql/alembic/versions/14346346596e_add_direction_type_column_in_route_table.py]: the is_forward is not in the database for the moment no ? (I think it's due to you previous modification)

**Comment by antoine-de** [comment]
@vpassama I merged you PR on navitia-proto, you can update the navitia-proto's SHA1 for jenkins to be able to build this PR

File: source/ed/connectors/fusio_parser.h (hunk 3 above)
Comment (vpassama) [inline source/ed/connectors/fusio_parser.h]: No it is not required, I fix this. About the eitri yes I did pip install and I had some issue with future.utils module  so I added "future==0.13.1" in requirements.

File: source/sql/alembic/versions/14346346596e_add_direction_type_column_in_route_table.py (hunk 4 above)
Comment (vpassama) [inline source/sql/alembic/versions/14346346596e_add_direction_type_column_in_route_table.py]: Yes I will remove this.

**Comment by vpassama** [comment]
Ok @antoine-de , I fixed all the little mistakes and updated data_version & navitia-proto sha1. Now we can focus on docker problem if you confirm that everything is all right. I'm on IRC btw.

**Comment by antoine-de** [comment]
thanks for the PR and the tests @vpassama !

**Pull request merged**